crates/sui-node/src/lib.rs

```diff
@@ -200,68 +204,84 @@ impl SuiNode {
         Ok(node_one_cell.get().await)
     }
 
-    fn start_jwk_updater(epoch_store: Arc<AuthorityPerEpochStore>) {
+    fn start_jwk_updater(
+        epoch_store: Arc<AuthorityPerEpochStore>,
+        consensus_adapter: Arc<ConsensusAdapter>,
+    ) {
         let epoch = epoch_store.epoch();
-        tokio::task::spawn(
-            async move {
-                info!("Starting JWK updater task");
-                loop {
-                    let epoch_store_ = epoch_store.clone();
-                    let supported_providers = epoch_store
-                        .protocol_config()
-                        .zklogin_supported_providers()
-                        .iter()
-                        .map(|s| OIDCProvider::from_str(s).expect("Invalid provider string"))
-                        .collect::<Vec<_>>();
-                    let fetch_and_sleep = async move {
-                        // Update the JWK value in the authority server
-                        info!("fetching new JWKs");
-                        match Self::fetch_jwks(&supported_providers).await {
+        let supported_providers = epoch_store
+            .protocol_config()
+            .zklogin_supported_providers()
+            .iter()
+            .map(|s| OIDCProvider::from_str(s).expect("Invalid provider string"))
+            .collect::<Vec<_>>();
+
+        info!(
+            "Starting JWK updater tasks with supported providers: {:?}",
+            supported_providers
+        );
+
+        for p in supported_providers.into_iter() {
+            let epoch_store = epoch_store.clone();
+            let consensus_adapter = consensus_adapter.clone();
+            spawn_monitored_task!(epoch_store.clone().within_alive_epoch(
+                async move {
+                    // note: restart-safe de-duplication happens after consensus, this is
+                    // just best-effort to reduce unneeded submissions.
+                    let mut seen = HashSet::new();
+                    loop {
+                        info!("fetching JWK for provider {:?}", p);
+                        match Self::fetch_jwks(&p).await {
                             Err(e) => {
                                 warn!("Error when fetching JWK {:?}", e);
                                 // Retry in 30 seconds
                                 tokio::time::sleep(Duration::from_secs(30)).await;
```

Comment:
Not part of this PR, but I believe in a case of error this will wait here for 30 sec, but then the next execution [line will be 265](https://github.com/MystenLabs/sui/pull/13530/files#diff-3fb10736526598f38da2b909064220eb69d70debb261372b7f452525105aa1a2R265) where we wait again for 1 hour - based on the comment I understand that we wanted to re-fetch again the jwks within 30 secs, right ?

Reply:
Good catch, thanks!